File: 2022/runner.ex
# Copyright 2022 Google LLC
#
# Use of this source code is governed by an MIT-style
# license that can be found in the LICENSE file or at
# https://opensource.org/licenses/MIT.
#
# Runner for Elixir Advent of Code solutions.
# Usage: elixir -r runner.ex day0/day0.exs day0/input*.txt

defmodule Runner do
  defmodule Result do
    defstruct outcome: :unknown, got: "", want: "", time_usec: 0

    def ok?(%Result{outcome: :fail}), do: false
    def ok?(%Result{}), do: true

    @outcome_colors %{
      success: :light_green_background,
      fail: :light_red_background,
      unknown: :light_yellow_background,
      todo: :cyan_background
    }
    @outcome_signs %{success: "✅", fail: "❌", unknown: "❓", todo: "❗"}
    def message(result) do
      outcome = Atom.to_string(result.outcome) |> String.upcase()
      sign = @outcome_signs[result.outcome]
      background = @outcome_colors[result.outcome]

      msg =
        case result do
          %Result{outcome: :success, got: got} -> "got #{got}"
          %Result{outcome: :fail, got: got, want: want} -> "got #{got}, want #{want}"
          %Result{outcome: :unknown, got: got} -> "got #{got}"
          %Result{outcome: :todo, want: ""} -> "implement it"
          %Result{outcome: :todo, want: want} -> "implement it, want #{want}"
        end

      IO.ANSI.format([sign, " ", background, :black, outcome, :reset, " ", msg])
    end
  end

  @doc """
  Parses argv as a command line and runs part1 and part2 of daymodule on each
  file in argv, or standard input if no files are given or any file is "-".
  Returns true if all parts  match their expected values for all files,
  otherwise returns false.  If `--verbose` or `-v` is present in argv,
  additional information like filenames, expected values, and timing will be
  printed to standard error.

  ## Examples:

    # Run against multiple files, print extra output
    iex> Runner.main(Day0, ~w[--verbose input.example.txt input.actual.txt])
    # Run against standard input, no extra output
    iex> Runner.main(Day0, [])
  """
  @spec main(module, [String.t()]) :: boolean
  def main(daymodule, argv) do
    {args, files} = OptionParser.parse!(argv, strict: [verbose: :boolean], aliases: [v: :verbose])
    verbose = Keyword.get(args, :verbose, false)
    files = if Enum.empty?(files), do: ["-"], else: files
    Enum.all?(Enum.map(files, &run(daymodule, &1, verbose)))
  end

  @doc """
  Reads input from file and passes it to :part1 and :part2 functions in
  daymodule.  Prints results.  Also checks for a companion .expected file and
  prints "SUCCESS" if the result matches, FAIL if it does not, "UNKNOWN" if the
  expected value is not yet set, and "TODO" if the function returned :todo.
  Returns true if no expected match failed, false otherwise.

  ## Examples

    iex> Runner.run(Day0, "day0/input.example.txt")
  """
  @spec run(module, String.t(), boolean) :: boolean
  def run(daymodule, file, verbose \\ true) do
    input = read_lines(file)
    expected = read_expected(file)

    outcomes =
      for part <- [:part1, :part2] do
        if verbose, do: IO.puts(:stderr, "Running #{daymodule} #{part} on #{file}")
        res = run_part(daymodule, part, input, Map.get(expected, part, ""))

        try do
          IO.puts("#{part}: #{res.got}")
        rescue
          Protocol.UndefinedError -> IO.inspect(res, label: part)
        end

        if verbose do
          IO.puts(:stderr, Result.message(res))
          IO.puts(:stderr, "#{part} took #{format_usec(res.time_usec)} on #{file}")
          IO.puts(:stderr, String.duplicate("=", 40))
        end

        Result.ok?(res)
      end

    Enum.all?(outcomes)
  end

  @doc """
  Runs function part of daymodule with the given iput and returns a
  Runner.Result based on the function's output and whether it matches want.

  ## Examples
      iex> Runner.run_part(Day0, :part1, ["123", "foo"], "42")
  """
  @spec run_part(module, atom, [String.t()], String.t()) :: Result.t()
  def run_part(daymodule, part, input, want) do
    {usec, got} = :timer.tc(daymodule, part, [input])

    outcome =
      cond do
        to_string(got) == to_string(want) -> :success
        got == :todo -> :todo
        want == "" -> :unknown
        true -> :fail
      end

    %Result{outcome: outcome, got: got, want: want, time_usec: usec}
  end

  @doc """
  Reads a file and splits it into a list of lines without trailing line breaks.
  If file is "-" reads from stdio.
  """
  @spec read_lines(String.t()) :: [String.t()]
  def read_lines(file) do
    case file do
      "" -> raise ArgumentError, "empty filename"
      "-" -> IO.stream()
      _ -> File.stream!(file)
    end
    |> Stream.map(&String.trim_trailing(&1, "\n"))
    |> Enum.to_list()
  end

  def read_expected(input_file) do
    file = Path.rootname(input_file, ".txt") <> ".expected"

    if File.exists?(file) do
      for line <- read_lines(file), String.starts_with?(line, "part") do
        [part, value] = String.split(line, ~r/:\s*/, parts: 2)
        {String.to_atom(part), value}
      end
      |> Map.new()
    else
      %{}
    end
  end

  defp format_usec(0), do: "0μs"
  defp format_usec(usec) when usec < 1000, do: "#{usec}μs"
  defp format_usec(usec) when usec < 1_000_000, do: "#{usec / 1000}ms"
  defp format_usec(usec) when usec < 60_000_000, do: "#{Float.round(usec / 1_000_000, 3)}s"
  defp format_usec(usec), do: "#{div(usec, 60_000_000)}:#{rem(div(usec, 1_000_000), 60)}"
end

File: 2025/runner.lua
#!/usr/bin/env lua
-- Copyright 2025 Trevor Stone
--
-- Use of this source code is governed by an MIT-style
-- license that can be found in the LICENSE file or at
-- https://opensource.org/licenses/MIT.

-- Advent of Code runner script for Lua.  Solution programs should call
-- require('runner'), then construct a table with part1 and part2 methods and
-- pass that to Runner.run.
Runner = {
  verbose = false,
  signs = {success = "✅", fail = "❌", unknown = "❓", todo = "❗"},
}

--[[
Run a solution on all inputs given to the program.
Expects a table with 'part1' and 'part2' functions as fields.
If -v argument is present, prints status information to stderr.
]]--
function Runner.run(day)
  local files = {}
  for i, a in ipairs(arg) do
    if a == '-v' then
      Runner.verbose = true
    else
      files[#files+1] = a
    end
  end
  if #files == 0 then
    files = {'-'} -- stdin
  end
  local success = true
  for i, f in ipairs(files) do
    success = Runner.runfile(day, f) and success
  end
  return success
end

function Runner.runfile(day, fname)
  local lines = {}
  local f
  if fname == '-' then
    f = io.input()
  else
    f = io.open(fname, 'r')
  end
  if not f then
    error(string.format('could not open file %s', fname))
  end
  for line in f:lines() do
    lines[#lines+1] = line
  end
  if f ~= '-' then
    f:close()
  end
  local expected = {}
  if fname:sub(-4) == '.txt' then
    local found, expectedlines = pcall(io.lines, fname:sub(1, -4) .. 'expected')
    if found then
      for l in expectedlines do
        local part, val = l:match('(part%d):%s*(.*)')
        if part then
          expected[part] = val
        end
      end
    end
  end
  local p1 = Runner.runpart(day, 'part1', lines, expected['part1'], fname)
  local p2 = Runner.runpart(day, 'part2', lines, expected['part2'], fname)
  return not (p1 == 'fail' or p2 == 'fail')
end

local function format_elapsed(seconds)
  if seconds < 0.001 then
    return string.format('%dμs', seconds * 1000000 // 1)
  elseif seconds < 1 then
    return string.format('%dms', seconds * 1000 // 1)
  elseif seconds < 60 then
    return string.format('%.3fs', seconds)
  elseif seconds < 3600 then
    return string.format('%d:%02d', seconds // 60, seconds % 60)
  else
    return string.format('%d:%02d:%02d', seconds // 3600, (seconds % 3600) // 60, seconds % 60)
  end
end

function Runner.runpart(day, part, lines, expected, fname)
  if Runner.verbose then
    io.stderr:write(string.format('Running %s %s on %s: (%d lines)\n', day, part, fname, #lines))
  end
  local starttime = os.time() -- seconds resolution
  local startclock = os.clock() -- CPU time, not wall time
  local result = tostring(day[part](lines))
  local endclock = os.clock()
  local endtime = os.time()
  local difftime = os.difftime(endtime, starttime)
  if difftime < 60 and difftime < (endclock - startclock) then
    difftime = endclock - startclock
  end
  print(string.format('%s: %s', part, result))
  local status = 'unknown'
  if result == expected then
    status = 'success'
  elseif result == 'TODO' then
    status = 'todo'
  elseif expected and expected ~= '' then
    status = 'fail'
  end
  if Runner.verbose then
    local message = ''
    if status == 'success' or status == 'unknown' then
      message = string.format('got %s', result)
    elseif status == 'fail' then
      message = string.format('got %s, wanted %s', result, expected)
    elseif status == 'todo' then
      if not expected or expected == '' then
        message = string.format('implement it')
      else
        message = string.format('implement it, want %s', expected)
      end
    end
    io.stderr:write(string.format('%s %s %s\n', Runner.signs[status], status:upper(), message))
    io.stderr:write(string.format('%s took %s on %s\n', part, format_elapsed(difftime), fname))
    io.stderr:write(string.rep('=', 40) .. '\n')
  end
  return status
end

function Runner.is_this_main()
  local callerfile = debug.getinfo(2).short_src
  return arg[0] == callerfile
end
